File: backend/app/api/v1/workflow.py
from datetime import datetime
from flask import request as flask_request
from flask_restx import Namespace, Resource, fields
from ...extensions import db
from ...models import ProcurementRequest, Approval, ApprovalDecision, RequestStatus
from ...services.approval_matrix import get_approval_steps

api = Namespace('workflow', description='Workflow actions')

approve_action_model = api.model('ApproveAction', {
    'action': fields.String(enum=['approve', 'reject', 'clarify'], required=True),
    'comment': fields.String,
    'approver_id': fields.Integer(required=True),  # replace with JWT later
    'role': fields.String(required=True),
})

approval_needed_model = api.model('ApprovalNeeded', {
    'roles': fields.List(fields.String),
})


@api.route('/<int:request_id>/needed')
class NeededApprovalsResource(Resource):
    @api.marshal_list_with(approval_needed_model)
    def get(self, request_id: int):
        req = ProcurementRequest.query.get_or_404(request_id)
        steps = get_approval_steps(float(req.amount))
        return [{'roles': s.roles} for s in steps]


@api.route('/<int:request_id>/action')
class WorkflowActionResource(Resource):
    @api.expect(approve_action_model, validate=True)
    def post(self, request_id: int):
        payload = flask_request.get_json()
        req = ProcurementRequest.query.get_or_404(request_id)
        action = payload['action']
        comment = payload.get('comment')
        role = payload['role']
        approver_id = payload['approver_id']

        if action == 'approve':
            decision = ApprovalDecision.APPROVED
        elif action == 'reject':
            decision = ApprovalDecision.REJECTED
        elif action == 'clarify':
            decision = ApprovalDecision.CLARIFY
        else:
            return {'message': 'Invalid action'}, 400

        approval = Approval(
            request_id=req.id,
            approver_id=approver_id,
            role=role,
            decision=decision,
            comment=comment,
            decided_at=datetime.utcnow(),
        )
        db.session.add(approval)

        # Update request status
        if decision == ApprovalDecision.REJECTED:
            req.status = RequestStatus.REJECTED
        elif decision == ApprovalDecision.CLARIFY:
            req.status = RequestStatus.CLARIFICATION
        else:
            # Simplified progression; detailed step tracking can be added later
            req.status = RequestStatus.UNDER_REVIEW

        db.session.commit()
        return {'message': 'Action recorded', 'status': req.status}, 200

File: backend/routes/workflow.js
const express = require('express');
const router = express.Router();
const Request = require('../models/Request');

// Submit new request
router.post('/submit', async (req, res) => {
  try {
    const newRequest = new Request(req.body);
    await newRequest.save();
    res.status(201).send(newRequest);
  } catch (error) {
    res.status(400).send(error);
  }
});

// Get requests for approval (role-based)
router.get('/pending', async (req, res) => {
  try {
    const pendingRequests = await Request.find({ status: 'pending' });
    res.status(200).send(pendingRequests);
  } catch (error) {
    res.status(500).send(error);
  }
});

// Add clarification comment
router.post('/:id/comment', async (req, res) => {
  try {
    const request = await Request.findById(req.params.id);
    request.comments.push(req.body.comment);
    await request.save();
    res.status(200).send(request);
  } catch (error) {
    res.status(400).send(error);
  }
});

// Approve/Reject/Request Info
router.post('/:id/action', async (req, res) => {
  try {
    const { action } = req.body;
    let status;
    if (action === 'approve') {
      status = 'approved';
    } else if (action === 'reject') {
      status = 'rejected';
    } else {
      return res.status(400).send({ message: 'Invalid action' });
    }
    const request = await Request.findByIdAndUpdate(
      req.params.id,
      { status },
      { new: true }
    );
    res.status(200).send(request);
  } catch (error) {
    res.status(400).send(error);
  }
});

module.exports = router;